File: 000-099/034-Find First and Last Position of Element in Sorted Array.java
class Solution {
    public int[] searchRange(int[] nums, int target) {
        int first = 0;
        int last = 0;
        int mid = 0;
        int n = nums.length;
        
        int left = 0;
        int right = n - 1;
        while (left <= right) {
            mid = left + (right - left) / 2;
            if (nums[mid] >= target)
                right = mid - 1;
            else
                left = mid + 1;
        }
        if (right + 1 < n)
            first = nums[right + 1] == target ? right + 1 : -1;
        else
            first = -1;
        
        left = 0;
        right = n - 1;
        while (left <= right) {
            mid = left + (right - left) / 2;
            if (nums[mid] <= target)
                left = mid + 1;
            else
                right = mid - 1;
        }
        if (left > 0)
            last = nums[left - 1] == target ? left - 1 : -1;
        else
            last = -1;
        
        return new int[]{first, last};
    }
}

File: 000-099/034-Find-First-and-Last-Position-of-Element-in-Sorted-Array.cpp
// Time complexity: O(logn).
// Space complexity: O(1).
class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        int first = -1;
        int last = -1;

        int lo = 0;
        int hi = nums.size() - 1;
        while (lo <= hi) {
            int mid = lo + (hi - lo) / 2;
            if (nums[mid] == target) {
                first = mid;
                hi = mid - 1;
            } else if (nums[mid] < target) {
                lo = mid + 1;
            } else {
                hi = mid - 1;
            }
        }
        if (first == -1) {
            return {-1, -1};
        }
        
        lo = first;
        hi = nums.size() - 1;
        while (lo <= hi) {
            int mid = lo + (hi - lo) / 2;
            if (nums[mid] == target) {
                last = mid;
                lo = mid + 1;
            } else if (nums[mid] < target) {
                lo = mid + 1;
            } else {
                hi = mid - 1;
            }
        }
        return {first, last};
    }
};
